### ISSUE TITLE
make: *** No rule to make target `/*.go', needed by `_output/bin/deepcopy-gen'.  Stop.

### ISSUE THREAD

**Issue by lixiaobing10051267** [kubernetes/kubernetes#28987]
No problem until yesterday, now I tried again many times, including clone again, but problem exists:

root@ubuntu:/home/zte/share/k8s-clone/kubernetes# git commit
Checking that it builds... 
make: *** No rule to make target `/*.go', needed by `_output/bin/deepcopy-gen'.  Stop.

Checking for docs that need updating... make: *** No rule to make target `/*.go', needed by `_output/bin/deepcopy-gen'.  Stop.
!!! Error in hack/verify-munge-docs.sh:29
  'make -C "${KUBE_ROOT}/" WHAT=cmd/mungedocs' exited with status 2
Call stack:
  1: hack/verify-munge-docs.sh:29 main(...)
Exiting with status 1
ERROR!
Some docs are out of sync between CLI and markdown.
To regenerate docs, run:
  hack/update-munge-docs.sh

Checking for swagger type documentation that need updating... ERROR!
Swagger type documentation needs to be updated.
To regenerate the spec, run:
  hack/update-generated-swagger-docs.sh

  1: /home/zte/share/k8s-clone/kubernetes/hack/lib/golang.sh:587 kube::golang::build_binaries(...)
  2: hack/make-rules/build.sh:26 main(...)
Exiting with status 1
make: *** [all] Error 1
!!! Error in hack/verify-swagger-spec.sh:26
  'make -C "${KUBE_ROOT}" WHAT=cmd/kube-apiserver' exited with status 2
Call stack:
  1: hack/verify-swagger-spec.sh:26 main(...)
Exiting with status 1
ERROR!
Swagger spec needs to be updated.
To regenerate the spec, run:
  hack/update-swagger-spec.sh

Aborting commit

**Comment by lixiaobing10051267** [comment]
many people have this problem, how can it be fixed?

**Comment by bboreham** [comment]
I had a similar problem that I fixed by ensuring I was building in a directory `.../k8s/kubernetes`; such that `../../github.com/` was a valid path.  Building in `.../github.com/kubernetes/kubernetes` failed.

**Comment by bboreham** [comment]
This is the same problem as #28890; the generator in `Makefile.generated_files` assumes that you have valid Go development environment set up, and if you don't then it cannot find the dependencies and generates a rule like this:

````
_output/bin/conversion-gen: cmd/libs/go2idl/conversion-gen \
	cmd/libs/go2idl/conversion-gen/main.go \
	/*.go \
	/*.go
````

(to illustrate:
````
dev:~/foobar/kubernetes$ unset GOPATH
dev:~/foobar/kubernetes$ (echo -n "$(DEEPCOPY_GEN): ";                          \
	 DIRECT=$$(go list -e -f '{{.Dir}} {{.Dir}}/*.go'      \
	     ./cmd/libs/go2idl/deepcopy-gen);                  \
	 INDIRECT=$$(go list -e                                \
	     -f '{{range .Deps}}{{.}}{{"\n"}}{{end}}'          \
	     ./cmd/libs/go2idl/deepcopy-gen                    \
	     | grep "^$(PRJ_SRC_PATH)"                         \
	     | xargs go list -e -f '{{.Dir}} {{.Dir}}/*.go');  \
	 echo $$DIRECT $$INDIRECT | sed 's/ / \\\n\t/g';       \
	) | sed "s|$$(pwd -P)/||"
````
)

**Comment by dims** [comment]
@lixiaobing10051267 - The recommended way as @bboreham mentioned is here:
https://github.com/kubernetes/kubernetes/blob/master/docs/devel/development.md#clone-your-fork

Also, fyi @thockin is exploring ways to add back the old use case here:
https://github.com/kubernetes/kubernetes/pull/28903

So until then please try the steps in the first url

**Comment by dims** [comment]
Based on a conversation in slack. Here's a check list.

1) Do you have GOPATH set? Is it a single directory?
2) Have you checked out kubernetes under `$GOPATH/src/k8s.io`? (Cross check the directory name!)
3) Can you `cd $GOPATH/src/k8s.io/kubernetes` and run `make clean`?
4) Can you then run `make gen_deepcopy`?
5) Can you run `make gen_conversion`?

**Comment by thockin** [comment]
#28903 has a couple options for fixing this.

**Comment by dims** [comment]
@thockin one more issue that came up this morning was that someone had checked out under `/home/xyz123/go/src/github.com/k8s.io/kubernetes` and set GOPATH to `/home/xyz123/go`. Was not sure if #28903 would support that case. (cc @bboreham)

**Comment by thockin** [comment]
no.  The canonical Go import path is `k8s.io/kubernetes` - importing it as
any other path (and expecting GOPATH to work) will not work.  That's just
how Go works (or doesn't, depending on your perspective).

I am pushing on #28903 which will allow the build to run regardless of
GOPATH (fixing the "regression" I introduced).  BUt I also offered #29010
which would assert that GOPATH was valid.  I am not as convinced that we
should do the latter, but I think the former is correct.

TL;DR: I broke "build from anywhere" and I am trying to fix it. :)

**Comment by bboreham** [comment]
> the canonical Go import path is `k8s.io/kubernetes`

Agreed.  But it is not at all obvious how to get from the symptom here to realising that you have broken this rule.

Note the Octocat link from http://http://kubernetes.io/ goes to https://github.com/kubernetes/kubernetes; the obvious next step is to clone to $GOPATH/github.com/kubernetes/kubernetes.

**Comment by thockin** [comment]
https://github.com/kubernetes/kubernetes/issues/29014

**Comment by bboreham** [comment]
The failure comes before any Go builds run.

Expanded out, the line I quoted at https://github.com/kubernetes/kubernetes/issues/28987#issuecomment-232932492 does this:

````
$ go list -e -f '{{.Dir}} {{.Dir}}/*.go' k8s.io/kubernetes/cmd/libs/go2idl/args k8s.io/kubernetes/cmd/libs/go2idl/deepcopy-gen/generators
 /*.go
 /*.go
````

when you have fetched the code somewhere other than `$GOPATH/k8s.io/kubernetes`

**Comment by bapparabi** [comment]
can you tell me what guide you using to install go as i try to use 

make gmc
make: *** No rule to make target 'gmc'. Stop.

many guide not working
